modules/h2/src/main/java/org/gridgain/internal/h2/table/TableFilter.java

```diff
@@ -312,21 +312,60 @@ public void prepare() {
         if (index.getClass() == HashJoinIndex.class)
             ((HashJoinIndex)index).prepare(session, indexConditions);
 
+        boolean hashIndex = index.getIndexType().isHash();
+        boolean[] columnPresenceMask = null;
+        int[] columnIndexes = null;
+        int maxColumnIdx = -1;
+
         // forget all unused index conditions
         // the indexConditions list may be modified here
         for (int i = 0; i < indexConditions.size(); i++) {
             IndexCondition condition = indexConditions.get(i);
             if (!condition.isAlwaysFalse()) {
                 Column col = condition.getColumn();
                 if (col.getColumnId() >= 0) {
-                    if (index.getColumnIndex(col) < 0) {
+                    int columnIdx = index.getColumnIndex(condition.getColumn());
+
+                    if (columnIdx < 0) {
```

Comment:
When can this condition be true?

If the condition is true, `indexCondition.size()` will be changed.
Thus, if you are already created `columnIndexes[]` array, then the array will be larger than `indexConditions`.
Therefore, the array gap may lead to out-of-bound exception below,
when you will remove conditions with unused index columns.